// gulp-flowtype/lib/index.js
/* @flow weak */
'use strict';

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

var Q = require('q');
var fs = require('fs');
require('babel-polyfill');
var path = require('path');
var gutil = require('gulp-util');
var through = require('through2');
var flowBin = require('flow-bin');
var logSymbols = require('log-symbols');
var childProcess = require('child_process');
var chalk = require('chalk');
var reporter = require('flow-reporter');

/**
 * Flow check initialises a server per folder when run,
 * we can store these paths and kill them later if need be.
 */
var servers = [];
var passed = true;

/**
 * Wrap critical Flow exception into default Error json format
 */
function fatalError(stderr) {
  return {
    errors: [{
      message: [{
        path: '',
        code: 0,
        line: 0,
        start: 0,
        descr: stderr
      }]
    }]
  };
}

function optsToArgs(opts) {
  var args = [];

  if (opts.all) {
    args.push('--all');
  }
  if (opts.weak) {
    args.push('--weak');
  }
  if (opts.declarations) {
    args.push('--lib', opts.declarations);
  }

  return args;
}

function getFlowBin() {
  return process.env.FLOW_BIN || flowBin;
}

function executeFlow(_path, options) {
  var deferred = Q.defer();

  var opts = optsToArgs(options);

  var command = opts.length || options.killFlow ? function () {
    servers.push(path.dirname(_path));
    return 'check';
  }() : 'status';

  var args = [command].concat(_toConsumableArray(opts), ['/' + path.relative('/', _path), '--json']);

  var stream = childProcess.spawn(getFlowBin(), args);

  var dat = "";
  stream.stdout.on('data', function (data) {
    dat += data.toString();
  });

  stream.stdout.on('end', function () {
    var parsed;
    try {
      parsed = JSON.parse(dat);
    } catch (e) {
      parsed = fatalError(dat);
    }
    var result = {};

    // loop through errors in file
    result.errors = parsed.errors.filter(function (error) {
      var isCurrentFile = error.message[0].path === _path;
      var generalError = /(Fatal)/.test(error.message[0].descr);

      return isCurrentFile || generalError;
    });

    if (result.errors.length) {
      passed = false;

      var report = typeof options.reporter === 'undefined' ? reporter : options.reporter;
      report(result.errors);

      if (options.abort) {
        deferred.reject(new gutil.PluginError('gulp-flow', 'Flow failed'));
      } else {
        deferred.resolve();
      }
    } else {
      deferred.resolve();
    }
  });

  return deferred.promise;
}

function checkFlowConfigExist() {
  var deferred = Q.defer();
  var config = path.join(process.cwd(), '.flowconfig');
  fs.exists(config, function (exists) {
    if (exists) {
      deferred.resolve();
    } else {
      deferred.reject('Missing .flowconfig in the current working directory.');
    }
  });
  return deferred.promise;
}

function hasJsxPragma(contents) {
  return (/@flow\b/ig.test(contents)
  );
}

function isFileSuitable(file) {
  var deferred = Q.defer();
  if (file.isNull()) {
    deferred.reject();
  } else if (file.isStream()) {
    deferred.reject(new gutil.PluginError('gulp-flow', 'Stream content is not supported'));
  } else if (file.isBuffer()) {
    deferred.resolve();
  } else {
    deferred.reject();
  }
  return deferred.promise;
}

function killServers() {
  var defers = servers.map(function (_path) {
    var deferred = Q.defer();
    childProcess.execFile(getFlowBin(), ['stop'], {
      cwd: _path
    }, deferred.resolve);
    return deferred;
  });
  return Q.all(defers);
}

module.exports = function () {
  var options = arguments.length <= 0 || arguments[0] === undefined ? {} : arguments[0];

  options.beep = typeof options.beep !== 'undefined' ? options.beep : true;

  function Flow(file, enc, callback) {
    var _this = this;

    var _continue = function _continue() {
      _this.push(file);
      callback();
    };

    isFileSuitable(file).then(function () {
      var hasPragma = hasJsxPragma(file.contents.toString());
      if (options.all || hasPragma) {
        checkFlowConfigExist().then(function () {
          executeFlow(file.path, options).then(_continue, function (err) {
            _this.emit('error', err);
            callback();
          });
        }, function (msg) {
          console.log(logSymbols.warning + ' ' + msg);
          _continue();
        });
      } else {
        _continue();
      }
    }, function (err) {
      if (err) {
        _this.emit('error', err);
      }
      callback();
    });
  }

  return through.obj(Flow, function () {
    var _this2 = this;

    var end = function end() {
      _this2.emit('end');
      passed = true;
    };

    if (passed) {
      console.log(logSymbols.success + ' Flow has found 0 errors');
    } else if (options.beep) {
      gutil.beep();
    }

    if (options.killFlow) {
      if (servers.length) {
        killServers().done(end);
      } else {
        end();
      }
    } else {
      end();
    }
  });
};
